### PR TITLE
Add a SWIFT_STDLIB_HAS_DARWIN_LIBMALLOC flag to allow/disallow uses of Darwin libmalloc APIs (malloc_default_zone, malloc_zone_malloc, etc.)

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test stdlib with toolchain

File: include/swift/Runtime/Concurrent.h [hunk 1]
```diff
@@ -601,7 +601,8 @@ template <class ElemTy> struct ConcurrentReadableHashMap {
   /// Otherwise, just return the passed-in size, which is always valid even if
   /// not necessarily optimal.
   size_t goodSize(size_t size) {
-#if defined(__APPLE__) && defined(__MACH__)
+#if defined(__APPLE__) && defined(__MACH__) && \
+        defined(SWIFT_STDLIB_HAS_DARWIN_LIBMALLOC)
```
Comment (mikeash) [inline include/swift/Runtime/Concurrent.h:605]: You'll want this to just be `SWIFT_STDLIB_HAS_DARWIN_LIBMALLOC` (no `defined()` around it) to handle the case where it's `0`.

**Review by mikeash — CHANGES_REQUESTED** [review]

**Comment by kubamracek** [comment]
@swift-ci please test

**Review by mikeash — APPROVED** [review]

**Pull request merged**